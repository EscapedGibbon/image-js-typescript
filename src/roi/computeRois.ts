import { Roi } from './Roi';
import { RoiMapManager } from './RoiMapManager';

/**
 * Generate an array of ROIs based on an ROI map manager.
 *
 * @param roiMapManager - Roi map manager to use
 */
export function computeRois(roiMapManager: RoiMapManager): void {
  const map = roiMapManager.getMap();

  const whites = new Array(map.nbPositive);
  const blacks = new Array(map.nbNegative);
  for (let i = 0; i < map.nbPositive; i++) {
    whites[i] = {
      borderLengths: [],
      borderIDs: [],
      minRow: map.height,
      minColumn: map.width,
      maxRow: -1,
      maxColumn: -1,
      surface: 0,
      id: i + 1,
      minX: Infinity,
      minY: Infinity,
      maxX: -Infinity,
      maxY: -Infinity,
    };
  }

  for (let i = 0; i < map.nbNegative; i++) {
    blacks[i] = {
      borderLengths: [],
      borderIDs: [],
      minRow: map.height,
      minColumn: map.width,
      maxRow: -1,
      maxColumn: -1,
      surface: 0,
      id: -i - 1,
      minX: Infinity,
      minY: Infinity,
      maxX: -Infinity,
      maxY: -Infinity,
    };
  }

  for (let row = 0; row < map.height; row++) {
    for (let column = 0; column < map.width; column++) {
      let currentIndex = roiMapManager.getMapValue(column, row);

      let currentRoi;
      if (currentIndex < 0) {
        currentRoi = blacks[-currentIndex - 1];
      } else {
        currentRoi = whites[currentIndex - 1];
      }

      currentRoi.surface++;

      if (row < currentRoi.minRow) {
        currentRoi.minRow = row;
      }
      if (row > currentRoi.maxRow) {
        currentRoi.maxRow = row;
      }
      if (column < currentRoi.minColumn) {
        currentRoi.minColumn = column;
      }
      if (column > currentRoi.maxColumn) {
        currentRoi.maxColumn = column;
      }
    }
  }

  roiMapManager.whiteRois = new Array<Roi>(map.nbPositive);
  roiMapManager.blackRois = new Array<Roi>(map.nbNegative);

  for (let i = 0; i < map.nbNegative; i++) {
    let blackRoi = new Roi(map, -i);
    blackRoi.origin = { row: blacks[i].minRow, column: blacks[i].minColumn };
    blackRoi.width = blacks[i].maxColumn - blacks[i].minColumn + 1;
    blackRoi.height = blacks[i].maxRow - blacks[i].minRow + 1;
    blackRoi.surface = blacks[i].surface;
    blackRoi.id = blacks[i].id;
    //computes minX,minY,maxX,maxY for black surfaces
    for (let y = 0; y < map.height; y++) {
      for (let x = 0; x < map.width; x++) {
        let target = y * map.width + x;
        if (map.data[target] === blackRoi.id) {
          const roi = blackRoi;
          if (x < roi.minX) {
            roi.minX = x;
          }
          if (x > roi.maxX) {
            roi.maxX = x;
          }
          if (y < roi.minY) {
            roi.minY = y;
          }
          if (y > roi.maxY) {
            roi.maxY = y;
          }
        }
      }
    }
    roiMapManager.blackRois[i] = blackRoi;
  }

  for (let i = 0; i < map.nbPositive; i++) {
    let whiteRoi = new Roi(map, i);
    whiteRoi.origin = { row: whites[i].minRow, column: whites[i].minColumn };
    whiteRoi.width = whites[i].maxColumn - whites[i].minColumn + 1;
    whiteRoi.height = whites[i].maxRow - whites[i].minRow + 1;
    whiteRoi.surface = whites[i].surface;
    whiteRoi.id = whites[i].id;
    //computes minX,minY,maxX,maxY for white surfaces
    for (let y = 0; y < map.height; y++) {
      for (let x = 0; x < map.width; x++) {
        let target = y * map.width + x;
        if (map.data[target] === whiteRoi.id) {
          const roi = whiteRoi;
          if (x < roi.minX) {
            roi.minX = x;
          }
          if (x > roi.maxX) {
            roi.maxX = x;
          }
          if (y < roi.minY) {
            roi.minY = y;
          }
          if (y > roi.maxY) {
            roi.maxY = y;
          }
        }
      }
    }
    roiMapManager.whiteRois[i] = whiteRoi;
  }
}
