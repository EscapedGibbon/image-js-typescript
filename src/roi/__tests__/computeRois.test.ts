import { fromMask } from '..';
import { computeRois } from '../computeRois';

test('3x3 mask', () => {
  const mask = testUtils.createMask([
    [1, 1, 1],
    [1, 0, 0],
    [0, 0, 1],
  ]);
  const roiMapManager = fromMask(mask);
  computeRois(roiMapManager);
  expect(roiMapManager.whiteRois).toHaveLength(2);
  expect(roiMapManager.blackRois).toHaveLength(1);

  expect(roiMapManager.whiteRois).toMatchSnapshot();
  expect(roiMapManager.blackRois).toMatchSnapshot();
});
