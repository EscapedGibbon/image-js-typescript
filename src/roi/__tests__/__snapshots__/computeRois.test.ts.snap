// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`3x3 mask 1`] = `
[
  {
    "aspectRatio": 0.554700196225229,
    "borderIDs": [
      -1,
    ],
    "borderLengths": [
      3,
    ],
    "boxIDs": [
      -1,
    ],
    "eqpc": 2.256758334191025,
    "external": 7,
    "externalIDs": [
      -1,
    ],
    "feretDiameterMax": {
      "angle": -33.690067525979785,
      "length": 3.605551275463989,
      "points": [
        {
          "column": 0,
          "row": 2,
        },
        {
          "column": 3,
          "row": 0,
        },
      ],
    },
    "feretDiameterMin": {
      "angle": 180,
      "length": 1.9999999999999996,
      "points": [
        {
          "column": 3,
          "row": 0,
        },
        {
          "column": 0,
          "row": 2,
        },
      ],
    },
    "fillRatio": 1,
    "height": 2,
    "id": 1,
    "internalIDs": [
      1,
    ],
    "maxX": 2,
    "maxY": 1,
    "minX": 0,
    "minY": 0,
    "ped": 2.0643291761037372,
    "perimeter": 6.485281374238571,
    "perimeterInfo": {
      "four": 0,
      "one": 0,
      "three": 2,
      "two": 2,
    },
    "roundness": 0.39176601376466547,
    "solidity": 0.8,
    "sphericity": 1.0932163146821783,
    "surface": 4,
    "width": 3,
  },
  {
    "aspectRatio": 0.7071067811865475,
    "borderIDs": [
      -1,
    ],
    "borderLengths": [
      2,
    ],
    "boxIDs": [],
    "eqpc": 1.1283791670955126,
    "external": 2,
    "externalIDs": [
      -1,
    ],
    "feretDiameterMax": {
      "angle": 45.00000000000001,
      "length": 1.4142135623730951,
      "points": [
        {
          "column": 0,
          "row": 0,
        },
        {
          "column": 1,
          "row": 1,
        },
      ],
    },
    "feretDiameterMin": {
      "angle": 90,
      "length": 1,
      "points": [
        {
          "column": 0,
          "row": 0,
        },
        {
          "column": 1,
          "row": 0,
        },
      ],
    },
    "fillRatio": 1,
    "height": 1,
    "id": 2,
    "internalIDs": [
      2,
    ],
    "maxX": 2,
    "maxY": 2,
    "minX": 2,
    "minY": 2,
    "ped": 0.4501581580785531,
    "perimeter": 1.4142135623730951,
    "perimeterInfo": {
      "four": 0,
      "one": 0,
      "three": 0,
      "two": 1,
    },
    "roundness": 0.6366197723675812,
    "solidity": 1,
    "sphericity": 2.5066282746310002,
    "surface": 1,
    "width": 1,
  },
]
`;

exports[`3x3 mask 2`] = `
[
  {
    "aspectRatio": 0.5547001962252291,
    "borderIDs": [
      1,
      2,
    ],
    "borderLengths": [
      3,
      1,
    ],
    "boxIDs": [
      1,
      2,
    ],
    "eqpc": 2.256758334191025,
    "external": 8,
    "externalIDs": [
      1,
      2,
    ],
    "feretDiameterMax": {
      "angle": -33.690067525979785,
      "length": 3.605551275463989,
      "points": [
        {
          "column": 0,
          "row": 2,
        },
        {
          "column": 3,
          "row": 0,
        },
      ],
    },
    "feretDiameterMin": {
      "angle": 0,
      "length": 2,
      "points": [
        {
          "column": 0,
          "row": 2,
        },
        {
          "column": 3,
          "row": 0,
        },
      ],
    },
    "fillRatio": 1,
    "height": 2,
    "id": -1,
    "internalIDs": [
      -1,
    ],
    "maxX": 2,
    "maxY": 2,
    "minX": 0,
    "minY": 1,
    "ped": 1.8006326323142123,
    "perimeter": 5.656854249492381,
    "perimeterInfo": {
      "four": 0,
      "one": 1,
      "three": 1,
      "two": 2,
    },
    "roundness": 0.39176601376466547,
    "solidity": 0.8,
    "sphericity": 1.2533141373155001,
    "surface": 4,
    "width": 3,
  },
]
`;
